ArduPlane/commands_logic.pde

```diff
@@ -506,6 +540,67 @@ static bool verify_loiter_turns()
     return false;
 }
 
+static bool verify_loiter_to_alt() {
+    update_loiter();
+
+    //has target altitude been reached?
+    if (condition_value != 0) {
+        if (fabs(condition_value - current_loc.alt) < 500) {
+            //Only have to reach the altitude once -- that's why I need
+            //this global condition variable.
+            //
+            //This is in case of altitude oscillation when still trying
+            //to reach the target heading.
+            condition_value = 0;
+        } else {
+            return false;
+        }
+    }
+    
+    //has target heading been reached?
+    if (condition_value2 != 0) {
+        //Get the lat/lon of next Nav waypoint after this one:
+        AP_Mission::Mission_Command next_nav_cmd;
+        if (! mission.get_next_nav_cmd(mission.get_current_nav_index() + 1,
+                                       next_nav_cmd)) {
+            //no next waypoint to shoot for -- go ahead and break out of loiter
+            return true;        
+        } 
+
+        // Bearing in radians
+        float bearing = (radians( (float)(get_bearing_cd(current_loc,next_nav_cmd.content.location)/100.0) ));
+
+        // Calculate heading
+        float heading = 5000;
+        if (((Compass &) compass).read()) {
+            const Matrix3f &m = ahrs.get_dcm_matrix();
```

Comment:
we shouldn't be using the compass directly. We may not even have a compass, or it could be very inaccurate. Just use ahrs.yaw_sensor, or ground course